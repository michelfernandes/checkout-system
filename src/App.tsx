import * as React from 'react';
import './App.css';
import { ProductTable } from './components/ProductTable';
import { PromotionalRulesList } from './components/PromotionalRulesList';
import { CheckoutTable, CheckoutScan } from './components/Checkout';

export type Product = {
  id: number,
  name: string,
  type: string,
  price: number
}

function App() {
  
  
  const [products, setProducts] = React.useState<Product[]>([
    {
      id: 1,
      name: "Curry Sauce",
      type: "Food",
      price: 1.95,
    },
    {
      id: 2,
      name: "Pizza",
      type: "Food",
      price: 5.99,
    },
    {
      id: 3,
      name: "Menâ€™s T-Shirt ",
      type: "Clothing",
      price: 25.00,
    }
  ]);

  const [totalPrice, setTotalPrice] = React.useState(0);
  const [scannedProducts, setScannedProducts] = React.useState<Product[]>([]);

  return (
    <>
      <div style={{textAlign: 'center'}}>
        <h2>Checkout System</h2>
      </div>
      <div style={{display: 'flex'}}>
        <div className="section-container">
          <ProductTable products={products} setProducts={setProducts}/>
        </div>
        <div className="section-container">
            <PromotionalRulesList/>
        </div>
        <div className="section-container">
          <CheckoutScan products={products} scannedProducts={scannedProducts} setScannedProducts={setScannedProducts} setTotalPrice={setTotalPrice}/>
          <CheckoutTable scannedProducts={scannedProducts} totalPrice={totalPrice} />
        </div>
      </div>
    </>
  );
}

export default App;
