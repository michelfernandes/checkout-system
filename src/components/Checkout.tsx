import * as React from "react";
import { Product } from '../App';
import { CheckoutService } from "../service/CheckoutService";
import { PizzaPriceRule } from "../rules/PizzaPriceRule";
import { ClothingDiscountRule } from "../rules/ClothingDiscountRule";

const checkoutService = new CheckoutService([new PizzaPriceRule(),new ClothingDiscountRule()]);
interface CheckoutScanProps {
    products: Product[];
    scannedProducts: Product[];
    setScannedProducts: (products: Product[]) => void;
    setTotalPrice: (price: number) => void;
}

export const CheckoutScan: React.FC<CheckoutScanProps> = ({
    products,
    scannedProducts,
    setScannedProducts,
    setTotalPrice
}) => {

  const [selected, setSelected] = React.useState(1);

  const handleScanClick = () => {
    const newScannedProduct = products.find(t => t.id === selected);
    checkoutService.scan(newScannedProduct!);
    setScannedProducts([...scannedProducts, newScannedProduct!]);
  }

  React.useEffect(() => {
    setTotalPrice(checkoutService.totalPrice);
  },[scannedProducts,setTotalPrice])

  const handleSelectionChange = (event) => {
    const selectedId = event.target.value;
    setSelected(Number(selectedId));
  }

  return (
    <>
      <div className="section-caption">
        <span>Checkout</span>
      </div>
      <div>
      <select onChange={handleSelectionChange} className="product-selection" name="products" id="productsSelectionList">
      {products.map((data) => (
        <option value={data.id}>{data.name}</option>
        ))}
      </select>
      <button className="crud-button" onClick={handleScanClick}>Scan</button>
      </div>
    </>
    
  );
};

interface CheckoutTableProps{
  scannedProducts: Product[]
  totalPrice: number
}

export const CheckoutTable: React.FC<CheckoutTableProps> = ({
  scannedProducts,
  totalPrice
}) => {
  return (
    <>
    <div>
      <table className="crud-table" >
        <tr>
          <th className="crud-table__header-cell">Product</th>
          <th className="crud-table__header-cell">Price</th>
        </tr>
      {scannedProducts.map((data) => (
        <tr>
          <td className="crud-table__cell">{data.name}</td>
          <td className="crud-table__cell">{data.price}</td>
        </tr>
      ))}
      </table>
    </div>
    <div className="section-caption">
        Total price: {totalPrice}
    </div>
    </>
    
  );
};


