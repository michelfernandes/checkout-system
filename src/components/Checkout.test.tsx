import { render } from '@testing-library/react';
import { Product } from '../App';
import { CheckoutScan, CheckoutTable } from './Checkout';

test('Should match Checkout Scan snapshot', () => {
  const checkoutScan = render(<CheckoutScan products={[]} scannedProducts={[]} setScannedProducts={function (products: Product[]): void {} } 
  setTotalPrice={function (price: number): void {} } />);
  expect(checkoutScan).toMatchSnapshot();
});

test('Should match Checkout Table snapshot', () => {
  const checkoutTable = render(<CheckoutTable scannedProducts={[]} totalPrice={0} />);
  expect(checkoutTable).toMatchSnapshot();
});
