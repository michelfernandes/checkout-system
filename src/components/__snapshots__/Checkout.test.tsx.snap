// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should match Checkout Scan snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="section-caption"
      >
        <span>
          Checkout
        </span>
      </div>
      <div>
        <select
          class="product-selection"
          id="productsSelectionList"
          name="products"
        />
        <button
          class="crud-button"
        >
          Scan
        </button>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="section-caption"
    >
      <span>
        Checkout
      </span>
    </div>
    <div>
      <select
        class="product-selection"
        id="productsSelectionList"
        name="products"
      />
      <button
        class="crud-button"
      >
        Scan
      </button>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`Should match Checkout Table snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div>
        <table
          class="crud-table"
        >
          <tr>
            <th
              class="crud-table__header-cell"
            >
              Product
            </th>
            <th
              class="crud-table__header-cell"
            >
              Price
            </th>
          </tr>
        </table>
      </div>
      <div
        class="section-caption"
      >
        Total price: 
        0
      </div>
    </div>
  </body>,
  "container": <div>
    <div>
      <table
        class="crud-table"
      >
        <tr>
          <th
            class="crud-table__header-cell"
          >
            Product
          </th>
          <th
            class="crud-table__header-cell"
          >
            Price
          </th>
        </tr>
      </table>
    </div>
    <div
      class="section-caption"
    >
      Total price: 
      0
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
